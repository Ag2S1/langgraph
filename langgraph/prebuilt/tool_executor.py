import json
from typing import Any, List, Optional, Sequence, Union

from langchain_core.load.serializable import Serializable
from langchain_core.runnables import RunnableBinding, RunnableConfig, RunnableLambda
from langchain_core.tools import BaseTool
from langchain_core.messages import AIMessage

INVALID_TOOL_MSG_TEMPLATE = (
    "{requested_tool_name} is not a valid tool, "
    "try one of [{available_tool_names_str}]."
)


class ToolInvocationInterface:
    """Interface for invoking a tool"""

    tool: str
    tool_input: Union[str, dict]


class ToolInvocation(Serializable):
    """Information about how to invoke a tool."""

    tool: str
    """The name of the Tool to execute."""
    tool_input: Union[str, dict]
    """The input to pass in to the Tool."""
    id: Optional[str] = None
    """The tool invocation ID"""


def create_tool_invocations(message: AIMessage) -> List[ToolInvocation]:
    if message.additional_kwargs.get("function_call"):
        return [
            ToolInvocation(
                tool=message.additional_kwargs["function_call"]["name"],
                tool_input=json.loads(
                    message.additional_kwargs["function_call"]["arguments"]
                ),
            )
        ]
    if message.additional_kwargs.get("tool_calls"):
        return [
            ToolInvocation(
                tool=tool["function"]["name"],
                tool_input=json.loads(tool["function"]["arguments"]),
                id=tool["id"],
            )
            for tool in message.additional_kwargs.get("tool_calls")
            if tool["type"] == "function"
        ]
    return []


class ToolExecutor(RunnableBinding):
    tools: Sequence[BaseTool]
    tool_map: dict
    invalid_tool_msg_template: str

    def __init__(
        self,
        tools: Sequence[BaseTool],
        *,
        invalid_tool_msg_template: str = INVALID_TOOL_MSG_TEMPLATE,
        **kwargs: Any,
    ) -> None:
        bound = RunnableLambda(self._execute, afunc=self._aexecute)
        super().__init__(
            bound=bound,
            tools=tools,
            tool_map={t.name: t for t in tools},
            invalid_tool_msg_template=invalid_tool_msg_template,
            **kwargs,
        )

    def _execute(
        self, tool_invocation: ToolInvocationInterface, *, config: RunnableConfig
    ) -> Any:
        if tool_invocation.tool not in self.tool_map:
            return self.invalid_tool_msg_template.format(
                requested_tool_name=tool_invocation.tool,
                available_tool_names_str=", ".join([t.name for t in self.tools]),
            )
        else:
            tool = self.tool_map[tool_invocation.tool]
            output = tool.invoke(tool_invocation.tool_input, config=config)
            return output

    async def _aexecute(
        self, tool_invocation: ToolInvocationInterface, *, config: RunnableConfig
    ) -> Any:
        if tool_invocation.tool not in self.tool_map:
            return self.invalid_tool_msg_template.format(
                requested_tool_name=tool_invocation.tool,
                available_tool_names_str=", ".join([t.name for t in self.tools]),
            )
        else:
            tool = self.tool_map[tool_invocation.tool]
            output = await tool.ainvoke(tool_invocation.tool_input, config=config)
            return output
