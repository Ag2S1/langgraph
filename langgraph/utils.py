import enum
from typing import Any, Awaitable, Callable, Optional

from langchain_core.runnables import Runnable, RunnableConfig
from langchain_core.runnables.config import merge_configs


# Before Python 3.11 native StrEnum is not available
class StrEnum(str, enum.Enum):
    """A string enum."""

    pass


class RunnableCallable(Runnable):
    """A much simpler version of RunnableLambda that requires sync and async functions."""

    def __init__(
        self,
        func: Callable[..., Optional[Runnable]],
        afunc: Optional[Callable[..., Awaitable[Optional[Runnable]]]] = None,
        *,
        name: Optional[str] = None,
        tags: Optional[list[str]] = None,
        trace: bool = True,
        **kwargs: Any,
    ) -> None:
        self.name = name or func.__name__
        self.func = func
        self.afunc = afunc
        self.config: Optional[RunnableConfig] = {"tags": tags} if tags else None
        self.kwargs = kwargs
        self.trace = trace

    def __repr__(self) -> str:
        repr_args = {
            k: v
            for k, v in self.__dict__.items()
            if k not in {"name", "func", "afunc", "config", "kwargs", "trace"}
        }
        return f"{self.get_name()}({', '.join(f'{k}={v!r}' for k, v in repr_args.items())})"

    def invoke(
        self, input: Any, config: Optional[RunnableConfig] = None, **kwargs: Any
    ) -> Any:
        kwargs = {**self.kwargs, **kwargs}
        if self.trace:
            ret = self._call_with_config(
                self.func, input, merge_configs(self.config, config), **kwargs
            )
        else:
            ret = self.func(input, merge_configs(self.config, config), **kwargs)
        if isinstance(ret, Runnable):
            return ret.invoke(input, config)
        return ret

    async def ainvoke(
        self, input: Any, config: Optional[RunnableConfig] = None, **kwargs: Any
    ) -> Any:
        if not self.afunc:
            return self.invoke(input, config, **kwargs)
        if self.trace:
            kwargs = {**self.kwargs, **kwargs}
            ret = await self._acall_with_config(
                self.afunc, input, merge_configs(self.config, config), **kwargs
            )
        else:
            kwargs = {**self.kwargs, **kwargs}
            ret = await self.afunc(input, merge_configs(self.config, config), **kwargs)
        if isinstance(ret, Runnable):
            return await ret.ainvoke(input, config)
        return ret
