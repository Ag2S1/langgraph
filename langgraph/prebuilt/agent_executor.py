import operator
from typing import Annotated, Optional, Sequence, TypedDict, Union

from langchain_core.agents import AgentAction, AgentFinish
from langchain_core.messages import BaseMessage
from langchain_core.runnables import RunnableLambda

from langgraph.graph import END, StateGraph
from langgraph.prebuilt.tool_executor import ToolExecutor


def _get_agent_state(input_schema=None):
    if input_schema is None:

        class AgentState(TypedDict):
            # The input string
            input: str
            # The list of previous messages in the conversation
            chat_history: Sequence[BaseMessage]
            # The outcome of a given call to the agent
            # Needs `None` as a valid type, since this is what this will start as
            agent_outcome: Union[AgentAction, AgentFinish, None]
            # List of actions and corresponding observations
            # Here we annotate this with `operator.add` to indicate that operations to
            # this state should be ADDED to the existing values (not overwrite it)
            intermediate_steps: Annotated[list[tuple[AgentAction, str]], operator.add]
            # Maximum number of iterations
            max_iterations: Optional[int]
            # Iteration count, tracked to enable early stopping
            iteration_count: int
            # # Maximum wall time
            # max_execution_time: Optional[float]
            # # Start time
            # start_time: float

    else:

        class AgentState(input_schema):
            # The outcome of a given call to the agent
            # Needs `None` as a valid type, since this is what this will start as
            agent_outcome: Union[AgentAction, AgentFinish, None]
            # List of actions and corresponding observations
            # Here we annotate this with `operator.add` to indicate that operations to
            # this state should be ADDED to the existing values (not overwrite it)
            intermediate_steps: Annotated[list[tuple[AgentAction, str]], operator.add]
            # Maximum number of iterations
            max_iterations: Optional[int]
            # Iteration count, tracked to enable early stopping
            iteration_count: int
            # # Maximum wall time
            # max_execution_time: Optional[float]
            # # Start time
            # start_time: float

    return AgentState


def _should_abort(data) -> bool:
    """Check if exceeding max iterations."""
    return (
        data.get("max_iterations") is not None
        and data.get("iteration_count", 0) >= data["max_iterations"]
        # or (
        #     data.get("max_execution_time") is not None
        #     and time.time() - data["start_time"] >= data["max_execution_time"]
        # )
    )


def create_agent_executor(agent_runnable, tools, input_schema=None):
    if isinstance(tools, ToolExecutor):
        tool_executor = tools
    else:
        tool_executor = ToolExecutor(tools)

    state = _get_agent_state(input_schema)

    # Define logic that will be used to determine which conditional edge to go down

    def should_continue(data):
        # If the agent outcome is an AgentFinish, then we return `exit` string
        # This will be used when setting up the graph to define the flow
        if isinstance(data["agent_outcome"], AgentFinish):
            return "end"
        # If exceeding max iterations
        elif _should_abort(data):
            return "abort"
        # Otherwise, an AgentAction is returned
        # Here we return `continue` string
        # This will be used when setting up the graph to define the flow
        else:
            return "continue"

    def run_agent(data):
        agent_outcome = agent_runnable.invoke(data)
        return {"agent_outcome": agent_outcome}

    async def arun_agent(data):
        agent_outcome = await agent_runnable.ainvoke(data)
        return {"agent_outcome": agent_outcome}

    # Define the function to execute tools
    def execute_tools(data):
        # Get the most recent agent_outcome - this is the key added in the `agent` above
        iteration_count = data.get("iteration_count", 0) + 1
        agent_action = data["agent_outcome"]
        if isinstance(agent_action, list):
            output = tool_executor.batch(agent_action, return_exceptions=True)
            return {
                "intermediate_steps": [
                    (action, str(out)) for action, out in zip(agent_action, output)
                ],
                "iteration_count": iteration_count,
            }
        output = tool_executor.invoke(agent_action)
        return {
            "intermediate_steps": [(agent_action, str(output))],
            "iteration_count": iteration_count,
        }

    async def aexecute_tools(data):
        # Get the most recent agent_outcome - this is the key added in the `agent` above
        iteration_count = data.get("iteration_count", 0) + 1
        agent_action = data["agent_outcome"]
        if isinstance(agent_action, list):
            output = await tool_executor.abatch(agent_action, return_exceptions=True)
            return {
                "intermediate_steps": [
                    (action, str(out)) for action, out in zip(agent_action, output)
                ],
                "iteration_count": iteration_count,
            }
        output = await tool_executor.ainvoke(agent_action)
        return {
            "intermediate_steps": [(agent_action, str(output))],
            "iteration_count": iteration_count,
        }

    def abort(data):
        return {
            "agent_outcome": AgentFinish(
                return_values={"output": "Agent stopped due to max iterations."},
                log="Agent stopped due to max iterations.",
            )
        }

    # Define a new graph
    workflow = StateGraph(state)

    # Define the two nodes we will cycle between
    workflow.add_node("agent", RunnableLambda(run_agent, arun_agent))
    workflow.add_node("action", RunnableLambda(execute_tools, aexecute_tools))
    workflow.add_node("abort", RunnableLambda(abort))

    # Set the entrypoint as `agent`
    # This means that this node is the first one called
    workflow.set_entry_point("agent")

    # We now add a conditional edge
    workflow.add_conditional_edges(
        # First, we define the start node. We use `agent`.
        # This means these are the edges taken after the `agent` node is called.
        "agent",
        # Next, we pass in the function that will determine which node is called next.
        should_continue,
        # Finally we pass in a mapping.
        # The keys are strings, and the values are other nodes.
        # END is a special node marking that the graph should finish.
        # What will happen is we will call `should_continue`, and then the output of that
        # will be matched against the keys in this mapping.
        # Based on which one it matches, that node will then be called.
        {
            # If `tools`, then we call the tool node.
            "continue": "action",
            # If `abort`, then we call the abort node.
            "abort": "abort",
            # Otherwise we finish.
            "end": END,
        },
    )

    # We now add a normal edge from `tools` to `agent`.
    # This means that after `tools` is called, `agent` node is called next.
    workflow.add_edge("action", "agent")

    # End if we abort due to exceeding iteration or time constraints.
    workflow.add_edge("abort", END)

    # Finally, we compile it!
    # This compiles it into a LangChain Runnable,
    # meaning you can use it as you would any other runnable
    return workflow.compile()
