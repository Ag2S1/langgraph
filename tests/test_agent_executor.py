import json
from typing import List

from langchain_core.agents import AgentFinish
from langchain_core.messages import AIMessage, BaseMessage
from langchain_core.prompts import ChatPromptTemplate, MessagesPlaceholder
from langchain_core.runnables import Runnable, RunnablePassthrough
from langchain_core.tools import BaseTool
from langchain_core.utils.function_calling import convert_to_openai_tool
from syrupy import SnapshotAssertion

from langgraph.prebuilt.agent_executor import create_agent_executor


def _make_agent_runnable(
    tools: List[BaseTool], response_messages: List[BaseMessage]
) -> Runnable:
    """Make fake agent runnable."""
    from langchain.agents.format_scratchpad.openai_tools import (
        format_to_openai_tool_messages,
    )
    from langchain.agents.output_parsers.openai_tools import (
        OpenAIToolsAgentOutputParser,
    )
    from langchain.chat_models.fake import FakeMessagesListChatModel

    chat_model = FakeMessagesListChatModel(responses=response_messages)
    prompt = ChatPromptTemplate.from_messages(
        [
            ("system", "You are a helpful assistant"),
            MessagesPlaceholder(variable_name="chat_history", optional=True),
            ("user", "{input}"),
            MessagesPlaceholder(variable_name="agent_scratchpad"),
        ]
    )
    llm_with_tools = chat_model.bind(
        tools=[convert_to_openai_tool(tool) for tool in tools]
    )

    return (
        RunnablePassthrough.assign(
            agent_scratchpad=lambda x: format_to_openai_tool_messages(
                x["intermediate_steps"]
            )
        )
        | prompt
        | llm_with_tools
        | OpenAIToolsAgentOutputParser()
    )


def test_agent_executor(snapshot: SnapshotAssertion) -> None:
    from langchain.agents import tool

    @tool()
    def search_api(query: str) -> str:
        """Searches the API for the query."""
        return f"result for {query}"

    tools = [search_api]
    responses = [
        AIMessage(
            content="",
            additional_kwargs={
                "tool_calls": [
                    {
                        "id": "tool_call123",
                        "type": "function",
                        "function": {
                            "name": "search_api",
                            "arguments": json.dumps("query"),
                        },
                    }
                ]
            },
        ),
        AIMessage(
            content="",
            additional_kwargs={
                "tool_calls": [
                    {
                        "id": "tool_call234",
                        "type": "function",
                        "function": {
                            "name": "search_api",
                            "arguments": json.dumps("another one"),
                        },
                    }
                ]
            },
        ),
        AIMessage(
            content="",
            additional_kwargs={
                "tool_calls": [
                    {
                        "id": "tool_call345",
                        "type": "function",
                        "function": {
                            "name": "search_api",
                            "arguments": json.dumps("another one"),
                        },
                    }
                ]
            },
        ),
        AIMessage(content="answer"),
    ]
    agent_runnable = _make_agent_runnable(tools, responses)
    app = create_agent_executor(agent_runnable, tools)

    assert app.get_input_schema().schema_json() == snapshot
    assert app.get_output_schema().schema_json() == snapshot
    assert json.dumps(app.get_graph().to_json(), indent=2) == snapshot
    assert app.get_graph().draw_ascii() == snapshot

    result = app.invoke(
        {
            "input": "what is the weather in sf?",
            "chat_history": [],
            "max_iterations": None,
            "iteration_count": 0,
            # "max_execution_time": None,
            # "start_time": time.time()
        }
    )
    assert result["agent_outcome"] == AgentFinish(
        return_values={"output": "answer"}, log="answer"
    )
    result = app.invoke(
        {
            "input": "what is the weather in sf?",
            "chat_history": [],
            "max_iterations": 2,
            "iteration_count": 0,
            # "max_execution_time": None,
            # "start_time": time.time()
        }
    )
    assert result["agent_outcome"] == AgentFinish(
        return_values={"output": "Agent stopped due to max iterations."},
        log="Agent stopped due to max iterations.",
    )
