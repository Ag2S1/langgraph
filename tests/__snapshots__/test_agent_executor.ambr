# serializer version: 1
# name: test_agent_executor
  '{"title": "LangGraphInput", "$ref": "#/definitions/AgentState", "definitions": {"BaseMessage": {"title": "BaseMessage", "description": "Base abstract Message class.\\n\\nMessages are the inputs and outputs of ChatModels.", "type": "object", "properties": {"content": {"title": "Content", "anyOf": [{"type": "string"}, {"type": "array", "items": {"anyOf": [{"type": "string"}, {"type": "object"}]}}]}, "additional_kwargs": {"title": "Additional Kwargs", "type": "object"}, "type": {"title": "Type", "type": "string"}, "name": {"title": "Name", "type": "string"}, "id": {"title": "Id", "type": "string"}}, "required": ["content", "type"]}, "AgentAction": {"title": "AgentAction", "description": "A full description of an action for an ActionAgent to execute.", "type": "object", "properties": {"tool": {"title": "Tool", "type": "string"}, "tool_input": {"title": "Tool Input", "anyOf": [{"type": "string"}, {"type": "object"}]}, "log": {"title": "Log", "type": "string"}, "type": {"title": "Type", "default": "AgentAction", "enum": ["AgentAction"], "type": "string"}}, "required": ["tool", "tool_input", "log"]}, "AgentFinish": {"title": "AgentFinish", "description": "The final return value of an ActionAgent.", "type": "object", "properties": {"return_values": {"title": "Return Values", "type": "object"}, "log": {"title": "Log", "type": "string"}, "type": {"title": "Type", "default": "AgentFinish", "enum": ["AgentFinish"], "type": "string"}}, "required": ["return_values", "log"]}, "AgentState": {"title": "AgentState", "type": "object", "properties": {"input": {"title": "Input", "type": "string"}, "chat_history": {"title": "Chat History", "type": "array", "items": {"$ref": "#/definitions/BaseMessage"}}, "agent_outcome": {"title": "Agent Outcome", "anyOf": [{"$ref": "#/definitions/AgentAction"}, {"$ref": "#/definitions/AgentFinish"}]}, "intermediate_steps": {"title": "Intermediate Steps", "type": "array", "items": {"type": "array", "minItems": 2, "maxItems": 2, "items": [{"$ref": "#/definitions/AgentAction"}, {"type": "string"}]}}, "max_iterations": {"title": "Max Iterations", "type": "integer"}, "iteration_count": {"title": "Iteration Count", "type": "integer"}}, "required": ["input", "chat_history", "agent_outcome", "intermediate_steps", "max_iterations", "iteration_count"]}}}'
# ---
# name: test_agent_executor.1
  '{"title": "LangGraphOutput", "type": "object", "properties": {"input": {"title": "Input", "type": "string"}, "chat_history": {"title": "Chat History", "type": "array", "items": {"$ref": "#/definitions/BaseMessage"}}, "agent_outcome": {"title": "Agent Outcome", "anyOf": [{"$ref": "#/definitions/AgentAction"}, {"$ref": "#/definitions/AgentFinish"}]}, "intermediate_steps": {"title": "Intermediate Steps", "type": "array", "items": {"type": "array", "minItems": 2, "maxItems": 2, "items": [{"$ref": "#/definitions/AgentAction"}, {"type": "string"}]}}, "max_iterations": {"title": "Max Iterations", "type": "integer"}, "iteration_count": {"title": "Iteration Count", "type": "integer"}}, "definitions": {"BaseMessage": {"title": "BaseMessage", "description": "Base abstract Message class.\\n\\nMessages are the inputs and outputs of ChatModels.", "type": "object", "properties": {"content": {"title": "Content", "anyOf": [{"type": "string"}, {"type": "array", "items": {"anyOf": [{"type": "string"}, {"type": "object"}]}}]}, "additional_kwargs": {"title": "Additional Kwargs", "type": "object"}, "type": {"title": "Type", "type": "string"}, "name": {"title": "Name", "type": "string"}, "id": {"title": "Id", "type": "string"}}, "required": ["content", "type"]}, "AgentAction": {"title": "AgentAction", "description": "A full description of an action for an ActionAgent to execute.", "type": "object", "properties": {"tool": {"title": "Tool", "type": "string"}, "tool_input": {"title": "Tool Input", "anyOf": [{"type": "string"}, {"type": "object"}]}, "log": {"title": "Log", "type": "string"}, "type": {"title": "Type", "default": "AgentAction", "enum": ["AgentAction"], "type": "string"}}, "required": ["tool", "tool_input", "log"]}, "AgentFinish": {"title": "AgentFinish", "description": "The final return value of an ActionAgent.", "type": "object", "properties": {"return_values": {"title": "Return Values", "type": "object"}, "log": {"title": "Log", "type": "string"}, "type": {"title": "Type", "default": "AgentFinish", "enum": ["AgentFinish"], "type": "string"}}, "required": ["return_values", "log"]}}}'
# ---
# name: test_agent_executor.2
  '''
  {
    "nodes": [
      {
        "id": "__start__",
        "type": "schema",
        "data": {
          "title": "LangGraphInput",
          "$ref": "#/definitions/AgentState",
          "definitions": {
            "BaseMessage": {
              "title": "BaseMessage",
              "description": "Base abstract Message class.\n\nMessages are the inputs and outputs of ChatModels.",
              "type": "object",
              "properties": {
                "content": {
                  "title": "Content",
                  "anyOf": [
                    {
                      "type": "string"
                    },
                    {
                      "type": "array",
                      "items": {
                        "anyOf": [
                          {
                            "type": "string"
                          },
                          {
                            "type": "object"
                          }
                        ]
                      }
                    }
                  ]
                },
                "additional_kwargs": {
                  "title": "Additional Kwargs",
                  "type": "object"
                },
                "type": {
                  "title": "Type",
                  "type": "string"
                },
                "name": {
                  "title": "Name",
                  "type": "string"
                },
                "id": {
                  "title": "Id",
                  "type": "string"
                }
              },
              "required": [
                "content",
                "type"
              ]
            },
            "AgentAction": {
              "title": "AgentAction",
              "description": "A full description of an action for an ActionAgent to execute.",
              "type": "object",
              "properties": {
                "tool": {
                  "title": "Tool",
                  "type": "string"
                },
                "tool_input": {
                  "title": "Tool Input",
                  "anyOf": [
                    {
                      "type": "string"
                    },
                    {
                      "type": "object"
                    }
                  ]
                },
                "log": {
                  "title": "Log",
                  "type": "string"
                },
                "type": {
                  "title": "Type",
                  "default": "AgentAction",
                  "enum": [
                    "AgentAction"
                  ],
                  "type": "string"
                }
              },
              "required": [
                "tool",
                "tool_input",
                "log"
              ]
            },
            "AgentFinish": {
              "title": "AgentFinish",
              "description": "The final return value of an ActionAgent.",
              "type": "object",
              "properties": {
                "return_values": {
                  "title": "Return Values",
                  "type": "object"
                },
                "log": {
                  "title": "Log",
                  "type": "string"
                },
                "type": {
                  "title": "Type",
                  "default": "AgentFinish",
                  "enum": [
                    "AgentFinish"
                  ],
                  "type": "string"
                }
              },
              "required": [
                "return_values",
                "log"
              ]
            },
            "AgentState": {
              "title": "AgentState",
              "type": "object",
              "properties": {
                "input": {
                  "title": "Input",
                  "type": "string"
                },
                "chat_history": {
                  "title": "Chat History",
                  "type": "array",
                  "items": {
                    "$ref": "#/definitions/BaseMessage"
                  }
                },
                "agent_outcome": {
                  "title": "Agent Outcome",
                  "anyOf": [
                    {
                      "$ref": "#/definitions/AgentAction"
                    },
                    {
                      "$ref": "#/definitions/AgentFinish"
                    }
                  ]
                },
                "intermediate_steps": {
                  "title": "Intermediate Steps",
                  "type": "array",
                  "items": {
                    "type": "array",
                    "minItems": 2,
                    "maxItems": 2,
                    "items": [
                      {
                        "$ref": "#/definitions/AgentAction"
                      },
                      {
                        "type": "string"
                      }
                    ]
                  }
                },
                "max_iterations": {
                  "title": "Max Iterations",
                  "type": "integer"
                },
                "iteration_count": {
                  "title": "Iteration Count",
                  "type": "integer"
                }
              },
              "required": [
                "input",
                "chat_history",
                "agent_outcome",
                "intermediate_steps",
                "max_iterations",
                "iteration_count"
              ]
            }
          }
        }
      },
      {
        "id": "__end__",
        "type": "schema",
        "data": {
          "title": "LangGraphOutput",
          "type": "object",
          "properties": {
            "input": {
              "title": "Input",
              "type": "string"
            },
            "chat_history": {
              "title": "Chat History",
              "type": "array",
              "items": {
                "$ref": "#/definitions/BaseMessage"
              }
            },
            "agent_outcome": {
              "title": "Agent Outcome",
              "anyOf": [
                {
                  "$ref": "#/definitions/AgentAction"
                },
                {
                  "$ref": "#/definitions/AgentFinish"
                }
              ]
            },
            "intermediate_steps": {
              "title": "Intermediate Steps",
              "type": "array",
              "items": {
                "type": "array",
                "minItems": 2,
                "maxItems": 2,
                "items": [
                  {
                    "$ref": "#/definitions/AgentAction"
                  },
                  {
                    "type": "string"
                  }
                ]
              }
            },
            "max_iterations": {
              "title": "Max Iterations",
              "type": "integer"
            },
            "iteration_count": {
              "title": "Iteration Count",
              "type": "integer"
            }
          },
          "definitions": {
            "BaseMessage": {
              "title": "BaseMessage",
              "description": "Base abstract Message class.\n\nMessages are the inputs and outputs of ChatModels.",
              "type": "object",
              "properties": {
                "content": {
                  "title": "Content",
                  "anyOf": [
                    {
                      "type": "string"
                    },
                    {
                      "type": "array",
                      "items": {
                        "anyOf": [
                          {
                            "type": "string"
                          },
                          {
                            "type": "object"
                          }
                        ]
                      }
                    }
                  ]
                },
                "additional_kwargs": {
                  "title": "Additional Kwargs",
                  "type": "object"
                },
                "type": {
                  "title": "Type",
                  "type": "string"
                },
                "name": {
                  "title": "Name",
                  "type": "string"
                },
                "id": {
                  "title": "Id",
                  "type": "string"
                }
              },
              "required": [
                "content",
                "type"
              ]
            },
            "AgentAction": {
              "title": "AgentAction",
              "description": "A full description of an action for an ActionAgent to execute.",
              "type": "object",
              "properties": {
                "tool": {
                  "title": "Tool",
                  "type": "string"
                },
                "tool_input": {
                  "title": "Tool Input",
                  "anyOf": [
                    {
                      "type": "string"
                    },
                    {
                      "type": "object"
                    }
                  ]
                },
                "log": {
                  "title": "Log",
                  "type": "string"
                },
                "type": {
                  "title": "Type",
                  "default": "AgentAction",
                  "enum": [
                    "AgentAction"
                  ],
                  "type": "string"
                }
              },
              "required": [
                "tool",
                "tool_input",
                "log"
              ]
            },
            "AgentFinish": {
              "title": "AgentFinish",
              "description": "The final return value of an ActionAgent.",
              "type": "object",
              "properties": {
                "return_values": {
                  "title": "Return Values",
                  "type": "object"
                },
                "log": {
                  "title": "Log",
                  "type": "string"
                },
                "type": {
                  "title": "Type",
                  "default": "AgentFinish",
                  "enum": [
                    "AgentFinish"
                  ],
                  "type": "string"
                }
              },
              "required": [
                "return_values",
                "log"
              ]
            }
          }
        }
      },
      {
        "id": "agent",
        "type": "runnable",
        "data": {
          "id": [
            "langchain_core",
            "runnables",
            "base",
            "RunnableLambda"
          ],
          "name": "run_agent"
        }
      },
      {
        "id": "action",
        "type": "runnable",
        "data": {
          "id": [
            "langchain_core",
            "runnables",
            "base",
            "RunnableLambda"
          ],
          "name": "execute_tools"
        }
      },
      {
        "id": "abort",
        "type": "runnable",
        "data": {
          "id": [
            "langchain_core",
            "runnables",
            "base",
            "RunnableLambda"
          ],
          "name": "abort"
        }
      },
      {
        "id": "agent_should_continue",
        "type": "unknown",
        "data": "agent_should_continue"
      }
    ],
    "edges": [
      {
        "source": "__start__",
        "target": "agent"
      },
      {
        "source": "abort",
        "target": "__end__"
      },
      {
        "source": "action",
        "target": "agent"
      },
      {
        "source": "agent",
        "target": "agent_should_continue"
      },
      {
        "source": "agent_should_continue",
        "target": "action",
        "data": "continue"
      },
      {
        "source": "agent_should_continue",
        "target": "abort",
        "data": "abort"
      },
      {
        "source": "agent_should_continue",
        "target": "__end__",
        "data": "end"
      }
    ]
  }
  '''
# ---
# name: test_agent_executor.3
  '''
                       +-----------+                    
                       | __start__ |                    
                       +-----------+                    
                              *                         
                              *                         
                              *                         
                         +-------+                      
                         | agent |*                     
                        *+-------+ ***                  
                      **              ****              
                    **                    ***           
                  **                         ***        
    +-----------------------+                   **      
    | agent_should_continue |                    *      
    +-----------------------+*                   *      
           *         **       ******             *      
         **            *            ******       *      
        *               **                ***    *      
  +-------+               *                 +--------+  
  | abort |             **                  | action |  
  +-------+            *                    +--------+  
           *         **                                 
            **     **                                   
              *   *                                     
          +---------+                                   
          | __end__ |                                   
          +---------+                                   
  '''
# ---
